feat(media dir nav): navigation through sub directories and breadcrumbs

andrew-paterson opened the pull request:
This is just a small first step in attempting nested media directories (https://github.com/netlify/netlify-cms/issues/3240)

So far it only works with Github, and only navigation up and down directories with breadcrumbs is implemented. As I'm not too familiar with the codebase, it would help to have someone give it a cursory sanity check, before continuing too far on the current path.

My approach has been to add a `listDirs` function to `packages/netlify-cms-backend-github/src/API.ts` and then update the `getMedia()` function in `packages/netlify-cms-backend-github/src/implementation.tsx` to run both `listFiles()` and `listDirs()` and concatenate the result.

This means both the directories and files are passed to the `MediaLibrary` component in the `files` array. As the Github type is included with each object in the array, the `toTableData()` function in `packages/netlify-cms-core/src/components/MediaLibrary/MediaLibrary.js` sets the `isDirectory` prop for each file object, based on whether it is a `tree` or `blob`. The `MediaLibraryCard` component can display and behave accordingly.

I'm struggling to see if it would be wiser to pass the directories to the component as a separate array altogether. One issue is that a directory should not technically have a `displayURL` prop, but if ommitted the following error is thown:

`Warning: Failed prop type: The prop files[0].displayURL is marked as required in MediaLibrary, but its value is undefined`

I'm also not 100% sure if the way in which I've added the `updateMediaFolder()` action is correct. Lastly, in order to get the default media directory into the breadcrumbs component, I have ended up adding an action called `getDefaultMediaFolder()` to set a prop called `defaultMediaFolder` and pass that to the component. 

I'm guessing there is an easier way to get a prop from the site config into a component?

erezrokah commented:
Hi @andrew-paterson, thank you for this - great draft PR.
I believe the code can be simplified as we can retrieve all of the nested files and directories in a single request by passing a `recursive` flag to GitHub (see `depth: 100` argument to `listFiles`), then doing the filtering on the client side.

Then we only need to maintain the `currentFolder` state in the UI component.

See my commit to reflect those changes and please let me know what you think.

I haven't look into the UI first, but I hope my changes address your questions.

I think the next steps are to see how we handle file uploads and also editorial workflow. For the latter each entry can have its own media folder at the global level, collection level and even at the field level.

andrew-paterson commented:
* Uploads to nested directories are working
  * As I can't get the `currentMediaFolder` from the app config, I am passing it to the `selectMediaFolder` function in `packages/netlify-cms-core/src/reducers/entries.ts` as an argument. I'm not sure if there is a better way to do this.
  * I have updated the `selectMediaFolder` function to simply return the `currentMediaFolder` if it is present, thus ignoring any config settings.
* Selecting and deleting multiple assets is implemented
  * Folders with children cannot be selected. What about making this behaviour configurable? I would probably prefer to be able to delete a folder and its contents recursively as an end user.
  * I'm using `await` to delete batches of files in sequence. I tried doing it asynchronously first, but the second commit request errored.
* I have begun updating some styling to match the spec in #2557. I pretty quickly reached a point where it was clear that updating some global styling settings would be necessary to go much further in doing this. I just wanted to check in about this before starting to overhaul the guts of what is there. If that's fine, I'm happy to take on #2557 as part of this.
* Is there a preferred place to put utility functions like `readableFileSize` which I have currently left in `packages/netlify-cms-core/src/components/MediaLibrary/MediaLibraryCard.js`? I would like that to be available in other components, and I thought I'd ask if any conventions exist on where to put it.
* For custom `media_folder` at the collection level, I need to get the custom folder setting into the `MediaLibrary` component, in order to filter correctly here:

      const currentMediaFolder = this.state.currentMediaFolder || defaultMediaFolder;
      const currentDirFiles = files.filter(file => dirname(file.path) === currentMediaFolder);

    where the above could be something like:

      const customMediaFolder = ***Not sure how to get this for a collection or field***
      const currentMediaFolder = this.state.currentMediaFolder || customMediaFolder || defaultMediaFolder;
      const currentDirFiles = files.filter(file => dirname(file.path) === currentMediaFolder);

* I also need a bit more time to understand the editorial workflow, as I have not used that yet, and so I don't really know how it will affect the media browser.

* I added `max` as a size to `packages/netlify-cms-ui-default/src/Icon.js` which sets the width and height to 100%, so you can have SVG icons that fill the available space.

* In terms of the next steps, would the ability to upload multiple files at once appeal? I think it would be good to show a UI with a select button and a drag and drop, and then show the list of queued files and update the UI for each one as it starts and finishes uploading.

mildred commented:
Thank you very much for you work, I started using Netlify a short while ago and I am still in the process of publishing my first website managed this way, and the first issue that came was that there is no hierarchy in the media folder (the site is composed of many photo galleries). I just discovered that you are working on this. And indeed, the next thing is going to be the file upload where uploading a single file at a time is not very convenient.

andrew-paterson commented:
@erezrokah thank you for the refactor! A few updates:

- I have implemented a new way of deriving the current media folder in cases where the media library has loaded for the first time, and the user has not yet navigated the directory structure. Instead of falling back to the default media folder from config, which is incorrect for collections or fields, I derive it from the file paths.

- I have also implemented sub-directory navigation when the media library is defined on a collection or field, and have added a config setting for `disable_media_folder_navigation` which prevents this if true. Please consider this experimental, and feel free to change it. As an end-user, I would love to retain the ability to limit the media browser to one directory for a specific field.

- The media library component sets state.currentMediaFolder to null whenever the modal is closed. This prevents a user from being in their last location when they close it for one field and open it for another.

I think the next things should be folder creation and uploading multiple files, but I will wait for your feedback before I start on any of that.

andrew-paterson commented:
Thank you @erezrokah. I'm going to move on to selecting and uploading multiple files, unless you see a reason not to.

erezrokah commented:
> Thank you @erezrokah. I'm going to move on to selecting and uploading multiple files, unless you see a reason not to.

Hi @andrew-paterson, I think selecting multiple files should be handled in another PR.
This will help keep the scope of this one down (it's already quite big and still missing support for all backends).

See https://github.com/netlify/netlify-cms/issues/2243

andrew-paterson commented:
@erezrokah good point. In that case, adding support for the other backends is the only other thing I can think of.  Currently only have a Github account, so I'm not sure how easily I can test for all the other backends. Is there a way of handling this situation at present?

erezrokah commented:
> I can test for all the other backends. Is there a way of handling this situation at present?

Hi @andrew-paterson, I can look into it later on this week.

andrew-paterson commented:
One feature I have overlooked is folder creation. I'm going to look at that now. One note is that the UI can allow the user to create a folder, but it will only be saved to the repo if the user adds at least one file to it, and uploads it. This is because I know that you can't have empty tree paths in Github.

andrew-paterson commented:
@erezrokah I have added folder creation. It allows the creation of emp[ty folders, and the user can enter a new folder and upload images into it. The component recognises when there is another folder with the same name and the submit functionality is disabled.

There are a couple of caveats:

- I am just pushing a new object into the `props.files` in the MediaLibrary component. I'm not 100% sure on this, but I can't really see how else to do it.
- The placeholder for the input is currently hard coded in English, at `packages/netlify-cms-core/src/components/MediaLibrary/MediaLibraryTop.js` line 109. Not sure how to get translations implemented for a new phrase.

vnourdin commented on packages/netlify-cms-core/src/components/MediaLibrary/MediaLibraryTop.js:
```diff
@@ -101,6 +104,11 @@ function MediaLibraryTop({
         />
+        <MediaLibraryCreateFolder
+          onKeyDown={onCreateFolder}
+          placeholder='Create folder' // TODO multilingual
```
To allow translations you just have to use `t` and add the default value in the english index file. You can see an example in the PR #4938 :wink:
Based on the translation line 104, I think we should name this one this way:
```suggestion
          placeholder={t('mediaLibrary.mediaLibraryModal.createFolder')}
```

vnourdin commented on packages/netlify-cms-core/src/components/MediaLibrary/MediaLibraryCreateFolder-old.js:
```diff
@@ -0,0 +1,61 @@
+import React from 'react';
```
As it's named `-old` I suppose this file should be removed, right?

andrew-paterson replied:
Yes, sorry. I will commit that now.

andrew-paterson commented:
I can't think of anything else that needs to be done here, in terms of features. Please shout if I have missed anything.

andrew-paterson commented on dev-test/config.yml:
```diff
@@ -1,10 +1,11 @@
 backend:
-  name: test-repo
+  name: github
+  repo: andrew-paterson/netlify-cms-hugo-test-site
```
Done.

andrew-paterson commented:
@vnourdin @erezrokah anything else I need to do here?

andrew-paterson commented:
I've just seen a comment in issue [4965](https://github.com/netlify/netlify-cms/issues/4965) which mentions that multiple file deletions in one commit are possible. This PR implements deletion of multiple images with a separate commit for each one. Perhaps we could look at that?

erezrokah commented:
> I've just seen a comment in issue [4965](https://github.com/netlify/netlify-cms/issues/4965) which mentions that multiple file deletions in one commit are possible. This PR implements deletion of multiple images with a separate commit for each one. Perhaps we could look at that?

Good point. I'm still working on getting https://github.com/netlify/netlify-cms/pull/4960 merged and will follow up on this one after

andrew-paterson commented:
@erezrokah thank you. The code below in `packages/netlify-cms-core/src/backend.ts` seems like it could be fairly easily updated updated to allow multiple media deletions? Happy to look at that, but will wait for your go-ahead. 

```async deleteMedia(config: Config, path: string) {
    const user = (await this.currentUser()) as User;
    const commitMessage = commitMessageFormatter(
      'deleteMedia',
      config,
      {
        path,
        authorLogin: user.login,
        authorName: user.name,
      },
      user.useOpenAuthoring,
    );
    return this.implementation.deleteFiles([path], commitMessage);
  }```

erezrokah commented:
So I finally had some time to give this is a try.

I would like to scope down the PR to only add directory navigation to make it easier to review.

1. Can we remove the code that adds the image type and file size? That should be done in a separate PR. Also that addition is not centered 
![image](https://user-images.githubusercontent.com/26760571/111306096-7605b500-8660-11eb-95aa-496a11c4e1ba.png)

2. Can we remove the multiple selection code? I think it should be done in a separate PR. Also, to my understanding that code only lets you delete multiple images?

3. This doesn't seem to work in editorial workflow (also notice the `draft` label is hidden by the checkbox icon):

https://user-images.githubusercontent.com/26760571/111307241-d5b09000-8661-11eb-8634-cab2dadc33de.mov

4. This breaks our tests - I think since before the change when a user uploads an image it was auto selected/focused and now it isn't.

5. More on editorial workflow - one can't insert existing images: 

https://user-images.githubusercontent.com/26760571/111307957-c7af3f00-8662-11eb-9d36-da2b763df7e9.mov

I'm not sure, but maybe it would be easier to handle this feature as a part of https://github.com/netlify/netlify-cms/issues/4965?

We would need to refactor the way we handle media asset to make it more pluggable first.

andrew-paterson commented:
That makes sense. I have removed the checkboxes and multiple selection code, and I've also reverted the part under the image to what it was. The only features now in this PR are folder navigation, breadcrumbs, and create a new folder.

Looks to be working ok, but I haven't had time to test thoroughly yet.

andrew-paterson commented:
@erezrokah I have added another commit to address the issues with the editorial workflow. I have changed the pattern now, so that:

The `selectMediaFiles` function returns all the files in the `mediaLibrary` object, even if the entry (collection or field) has its own `media_folder` setting.

I have added a function called `getStartingMediaFolder` which returns the value of `media_folder` for a collection or entry, and if this exists, it is used as the `currentMediaFolder` value when the user opens the media library from the relevant field.

If `disable_media_folder_navigation` is not true for that field, then the user is still able to create new folders, and navigate up and down the folder tree, so the `media_folder` setting on the field or collection simply acts as a starting point.

However, if `disable_media_folder_navigation` is true, the breadcrumbs and folder cards become disabled, and the create folder input is hidden, thus forcing the user to use only the folder specified in the `media_folder` setting of the field or collection.

erezrokah commented:
Thanks @andrew-paterson, sorry for me being so slow on this. I've been clearing some of the other smaller PRs so I can focus  on this one

erezrokah commented:
It seems the global media library is broken.
You can see this by going to https://deploy-preview-4670--cms-demo.netlify.app/ and clicking `Media`.

> The `selectMediaFiles` function returns all the files in the `mediaLibrary` object, even if the entry (collection or field) has its own `media_folder` setting.

Can you explain the reason for this change? This breaks existing support for setting a media folder per collection/field.

I fixed the formatting, linting and types error, but some tests are still failing, I suspect some of them due to the `selectMediaFiles` change.

andrew-paterson commented:
My apologies, I was sure I had checked the global media library, but it should be fixed now.

To explain the change to `selectMediaFiles ` more fully. 

I have added a new function to `/reducers/mediaLibrary.ts` called `getStartingMediaFolder`. This returns `selectMediaFolder(state.config, collection, entry, field);` in any case where there is not an integration. My understanding is that this will return the field level `media_folder` if it exists, otherwise the collection level `media_folder` if that exists, or the global `media_folder`.

The media library compoent uses this code to then determine the current media folder:

`const currentMediaFolder = this.state.currentMediaFolder || startingMediaFolder;`

`this.state.currentMediaFolder` will return the user selected folder location if that exists. This would be from either clicking an item in the breadcrumbs, or a folder item in the media grid, shown here:

![image](https://user-images.githubusercontent.com/12248962/114002129-e4532700-985c-11eb-9b26-28a380c44f47.png)

If `this.state.currentMediaFolder` does not exist (ie the user has just opened the media library) the current location will be set to the result of `getStartingMediaFolder` referred to above. 

Therefore, if a field has a `media_folder` defined, when the user opens the media library from that field, it will start on that folder. However, because the `selectMediaFiles ` function has now returned everything in the media folder, he user will still be able to navigate up or down the folder structure. 

This can however be prevented by also setting `disable_media_folder_navigation: true` for that field in `config.yml`. In this case, the user will still be taken to the field `media_folder` location when opening the media library from thst field, but the navigation is disabled.

An example to illustrate.

My project has the following in the `static/media` directory:

![image](https://user-images.githubusercontent.com/12248962/114003431-06997480-985e-11eb-82a1-bf63bc065c0b.png)

Assume my global config has the following:

```media_folder: static/media
public_folder: /media```

Opening the global media library would give me this:

![image](https://user-images.githubusercontent.com/12248962/114003232-deaa1100-985d-11eb-957a-2634c1c9b599.png)

Assume my config for a field has `media_folder: /static/media/images`

When I open the media folder from that field by clicking "Choose image" I get:

![image](https://user-images.githubusercontent.com/12248962/114004353-ddc5af00-985e-11eb-9988-cb1a4785f979.png)
 
In this case, the user can still navigate by clicking the breadcrumbs or the folders.

Assume that for the same field I add `disable_media_folder_navigation: true` to the config.

When I open the media folder from that field by clicking "Choose image" I get:

![image](https://user-images.githubusercontent.com/12248962/114004717-3ac16500-985f-11eb-931b-e3875ca4cdfc.png)

It looks similar, but the folders and breadcrumbs are not clickable. (I need a better way of indicating that the folders are not clickable).

erezrokah commented:
> My apologies, I was sure I had checked the global media library, but it should be fixed now.

I don't see this change committed to the PR. Can you make you've pushed your changes?

> To explain the change to `selectMediaFiles ` more fully.

Thank you for the explanation. That makes sense when not working with editorial workflow.
However with editorial workflow this doesn't work. When editing a draft entry we save media files on the entry state so we can commit them together in a single commit.
Also, loading all media files when editing a draft can lead to collisions.
For the following configuration and directory structure:
```yaml
backend:
  name: github
  branch: netlify_cms/pr_4670
  repo: erezrokah/netlify-cms-reproductions

publish_mode: editorial_workflow
media_folder: static/media
collections:
  - name: posts
    label: Posts
    label_singular: 'Post'
    folder: content/posts
    create: true
    media_folder: /static/media/collection
    fields:
      - label: Title
        name: title
        widget: string
      - label: 'Cover Image 1'
        name: 'image1'
        widget: 'image'
        required: false
        media_folder: ''
      - label: 'Cover Image 2'
        name: 'image2'
        widget: 'image'
        required: false
      - label: 'Cover Image 3'
        name: 'image3'
        widget: 'image'
        media_folder: /static/media/collection/image3
        required: false
      - label: 'Cover Image 4'
        name: 'image4'
        widget: 'image'
        media_folder: /other/media/image4
        required: false
```

```
static
└── media
    ├── collection
    │   └── netlify.png
    ├── netlify.png
    └── netlify_2.png
other
└── media
    └── image4
        └── netlify.png
```

see the following video:

https://user-images.githubusercontent.com/26760571/114012818-7c5a0c00-986f-11eb-9d8f-7f2456288b2e.mov

Also, navigation doesn't work when the field has a `media_folder` which is not a prefix of the global media folder.
See behavior for `image4` field:

https://user-images.githubusercontent.com/26760571/114013139-e1adfd00-986f-11eb-8653-882d5ef0bb8c.mov

Would it make sense to scope this down to only the global media folder and not when editing entries?
I feel that directory navigation is more useful for the global one.

andrew-paterson commented:
My apologies for the push issue, it should be there now. I will check on the above issues with the editorial workflow and revert.

My initial impression on scoping this down to the global media library could be a good option- will look into that as well.

andrew-paterson commented:
@erezrokah I've begun scoping this feature down. I've reverted the `selectMediaFiles` function locally, and it seems that has fixed the issues with the editorial workflow, and the folder contents load fine when the collection of field has a folder defined, but the ability to navigate is disabled. 

I think this is fine where a `media_folder` is set on a field or collection. However, I would like the media library to still be navigable f there is not a `media_folder` defined on the field or collection- in other words, I want it to then behave exactly like the global media library.

However, on both master and in this branch, if the only `media_folder` setting is in the global level config, then the result of this code in the `selectMediaFiles` is an empty array, and the media library is empty.

```
const entryFiles = entryDraft
      .getIn(['entry', 'mediaFiles'], List<MediaFileMap>())
      .toJS() as MediaFile[];
```
This code only runs when `editingDraft` is true. Therefore, when the global media library is opened, this code is run:

`mediaLibrary.get('files') || [];`

and you get all the files inside the global `media_folder` directory.

That was why I had originally changed `selectMediaFiles` to just return `mediaLibrary.get('files') || []` in all cases.

Is it by design that the media library is empty when opened from a field, where neither the field nor collection has a `media_folder` setting?

For reference, here is my config:

```
media_folder: /static/media/
collections:
  - name: posts
    label: Posts
    label_singular: 'Post'
    folder: content/posts
    create: true
    editor:
      preview: false
    fields:
      - label: Title
        name: title
        widget: string
      - label: 'Cover Image 1'
        name: 'image1'
        widget: 'image'
        required: false
      - label: 'Cover Image 2'
        name: 'image2'
        widget: 'image'
        required: false
      - label: 'Cover Image 3'
        name: 'image3'
        widget: 'image'
        required: false
      - label: 'Cover Image 4'
        name: 'image4'
        widget: 'image'
        required: false
```

erezrokah commented:
Hi @andrew-paterson, thank you for making those changes.
I formatted the files and fixed some lint/type errors. Some tests are still failing, can you look into it?

andrew-paterson commented:
@erezrokah thank you. I have pushed an update where inserting files from a field seems to be working better now, both where a field has a `media_folder` setting and where it does not (In this case, the navigation is enabled). Things are looking much better now after a lot of manual testing.

Which automated tests are you running locally to check this? There are multiple listed [here](https://github.com/netlify/netlify-cms/blob/master/CONTRIBUTING.md), and I just want to make sure we're on the same page.

andrew-paterson commented:
Running tests for http://localhost:8080/__/#/tests/integration/media_library_spec_github_backend_rest.js, I get a failure when trying to save the entry. API 404 seems to be an error from the mock server. Is this correct? I just want to make sure I understand where this error is coming from. 

Incidentally, saving an entry with an image using the editorial workflow works fine when I do it manually, pointing to my own Github repo.

<img width="1438" alt="image" src="https://user-images.githubusercontent.com/12248962/115366404-9aa3fe00-a1c5-11eb-876d-2a0b925de3fb.png">

erezrokah commented:
> API 404 seems to be an error from the mock server. Is this correct? I just want to make sure I understand where this error is coming from.

Yes the server is returning 404 for a specific request. We use pre-recorded data, so something might have changed with the requests the CMS sends.

I would first make sure the non GitHub/Bitbucket/GitLab/Git Gateway tests work as they don't rely on pre-recorded data.

andrew-paterson commented:
Ok all of the following tests are passing:

```editorial_workflow_spec_test_backend.js
field_validations_spec.js
i18n_editorial_workflow_spec_test_backend.js
markdown_widget_backspace_spec.js
markdown_widget_code_block_spec.js
markdown_widget_enter_spec.js
markdown_widget_hotkeys_spec.js
markdown_widget_link_spec.js
markdown_widget_list_spec.js
markdown_widget_marks_spec.js
markdown_widget_quote_spec.js
media_library_spec_test_backend.js
search_suggestion_spec.js
simple_workflow_spec_test_backend.js
view_filters_spec.js
view_groups_spec.js```

I believe those are all the e2e tests that don't rely on a mock backend service. Any ideas on how to approach the backend specific tests? I have no idea what is going wrong with those.

erezrokah commented:
> I believe those are all the e2e tests that don't rely on a mock backend service. Any ideas on how to approach the backend specific tests? I have no idea what is going wrong with those.

Let me look into the Cypress failing tests. Does `yarn test` pass on your machine (these are unit tests)? It's failing on CI at the moment.

andrew-paterson commented:
Let me just push quickly.

andrew-paterson commented:
The result of `yarn test`. Seems like there are linting errors in two files if I'm reading that correctly?
```
~/Documents/development/netlify-cms [feat/nested_media_dirs] $ yarn test
yarn run v1.9.4
$ run-s clean lint type-check test:unit
$ rimraf "packages/*/dist" dev-test/dist
$ run-p -c --aggregate-output "lint:*"
$ stylelint --ignore-path .gitignore "{packages/**/*.{css,js,jsx,ts,tsx},website/**/*.css}"
Build Package: netlify-cms
error Command failed with exit code 1.
$ prettier "{{packages,scripts,website}/**/,}*.{js,jsx,ts,tsx,css}" --list-different
packages/netlify-cms-core/src/actions/mediaLibrary.ts
packages/netlify-cms-core/src/reducers/entries.ts
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
$ eslint --color --ignore-path .gitignore "{{packages,scripts,website}/**/,}*.{js,jsx,ts,tsx}"
Build Package: netlify-cms
ERROR: "lint:format" exited with 1.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
ERROR: "lint" exited with 1.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

erezrokah commented:
You can run `yarn format` to fix the formatting issues, then run `yarn test` again

andrew-paterson commented:
Thanks. I need to do that and push one more update.

andrew-paterson commented:
Ok there are 4 unit tests failing, that I'm struggling with a little- notes below.

The same e2e tests mentioned above are all passing.

## 1

`entries › selectMediaFilePublicPath › should resolve path from public folder for collection with no media folder`

The `selectMediaFilePublicPath` now removed the `media_folder` from the beginning of the path and replaces it with the `public_folder`, so that the full path of the image be maintained. It was previously just joining the public folder and the filename.

This test only passes the public folder to the function:

```
selectMediaFilePublicPath(
  { public_folder: '/media' },
  null,
  '/media/image.png',
  undefined,
  undefined,
),
```

I don't think this is allowed by the actual application, and I'm not sure I can get this test to pass with the new pattern.

## 2 

```
FAIL  packages/netlify-cms-core/src/actions/__tests__/mediaLibrary.spec.js  
  ● mediaLibrary › persistMedia › should not persist media when editing draft
```

```
FAIL  packages/netlify-cms-core/src/reducers/__tests__/mediaLibrary.spec.js
  ● mediaLibrary › should select draft media files from field when editing a draft
  ● mediaLibrary › should select draft media files from collection when editing a draft
```

In the latest version of the `selectMediaFiles ` function, if `(editingDraft && !integration)` is true, I get the enry files, and concatenate them with `mediaLibrary.get('files')`, as you need all the files to allow directory navigation when the media library is opened from a field. Previously this function would only return the entry files.

It's working fine for me in the app, but in the unit tests, `mediaLibrary.get('files')` is returning an empty list:

```
List {
  size: 0,
  _origin: 0,
  _capacity: 0,
  _level: 5,
  _root: undefined,
  _tail: undefined,
  __ownerID: undefined,
  __hash: undefined,
  __altered: false
}
```
After concatenating the media files and entry files, I don't get any objects, and this is causing issues.

I'm not sure if this is just an issue with the tests themselves, or if this is a problem for the wider application?

SignSpice commented:
Curious what the status is on this pull request. I am very interested in this feature.

SignSpice commented:
@erezrokah What still needs to be done for this to be merged? Is it only the two merge conflicts and the 3 (I think) failing tests?

erezrokah commented:
> @erezrokah What still needs to be done for this to be merged? Is it only the two merge conflicts and the 3 (I think) failing tests?

Hi @SignSpice

1. Fix merge conflicts
2. Look into existing tests failures and verify it doesn't break current behavior (specifically nested collections, global media folder, collection level media folder and field level media folder)
3. Add support for all backends
4. Add new tests (e.g Cypress)
5. More thorough code review

martinjagodic commented:
@andrew-paterson are you still interested in moving this forward?
